src/Avalonia.X11/X11WindowModes/XEmbedClientWindowMode.cs

```diff
@@ -0,0 +1,179 @@
+#nullable enable
+using System;
+using Avalonia.Controls;
+using Avalonia.Controls.Embedding;
+using Avalonia.Input;
+
+namespace Avalonia.X11;
+using static XLib;
+partial class X11Window
+{
+    public class XEmbedClientWindowMode : X11WindowMode
+    {
+        EmbeddableControlRoot? Root => Window._inputRoot as EmbeddableControlRoot;
+        private bool _focusedInEmbedder;
+        private bool _embedderActivated;
+        private IInputElement? _savedFocus;
+        private bool _disabled;
+        private IntPtr _currentEmbedder;
+        private bool _suppressConfigureEvents;
+        
+        public override bool BlockInput => _disabled;
+        public double Scaling
+        {
+            get => Window._scalingOverride ?? 1;
+            set => Window._scalingOverride = value;
+        }
+
+        public override void OnHandleCreated(IntPtr handle)
+        {
+            var data = new[]
+            {
+                IntPtr.Zero, new(1) /* XEMBED_MAPPED */
+            };
+
+            XChangeProperty(Display, handle, X11.Atoms._XEMBED_INFO, X11.Atoms._XEMBED_INFO, 32, PropertyMode.Replace,
+                data, data.Length);
+            Scaling = 1;
+            
+            base.OnHandleCreated(handle);
+        }
+
+        void SendXEmbedMessage(XEmbedMessage message, IntPtr detail = default, IntPtr data1 = default, IntPtr data2 = default)
+        {
+            if (_currentEmbedder == IntPtr.Zero)
+                return;
+            var xev = new XEvent
+            {
+                ClientMessageEvent =
+                {
+                    type = XEventName.ClientMessage,
+                    send_event = 1,
+                    window = _currentEmbedder,
+                    message_type = X11.Atoms._XEMBED,
+                    format = 32,
+                    ptr1 = default,
+                    ptr2 = new ((int)message),
+                    ptr3 = detail,
+                    ptr4 = data1,
+                    ptr5 = data2
+                }
+            };
+            XSendEvent(X11.Display, _currentEmbedder, false,
+                new IntPtr((int)(EventMask.NoEventMask)), ref xev);
+        }
+        
+        static XEmbedClientWindowMode()
+        {
+            KeyboardDevice.Instance.PropertyChanged += (_, args) =>
```

Comment:
We don't have a good stable API for that. Needs something like [FocusManager.GotFocus](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.input.focusmanager.gotfocus?view=winrt-26100).
But `InputElemement.GotFocusEvent.AddClassHandler` should work here.

Reply:
IInputElement isn't necessary implemented by InputElement, there are plans for focusable non-visual-based input elements for text views (like in WPF)

Reply:
Pretty sure these still should raise GotFocusEvent, even not inheriting InputElement. But it can be revisited later.